[MPS] Fix addmm

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116548
* __->__ #116547

Remove weird logic for designating matrices as transposed if sizes match(which always true if square matrices are multiplied with each other), which resulted in `torch.addmm` returns transposed matrix compared to `torch.mm`, see below:
```
% python -c "import torch;torch.set_default_device('mps');a=torch.eye(2);b=torch.arange(4.0).reshape(2, 2);print(a@b);print(torch.addmm(torch.zeros(2, 2), a,b))"
tensor([[0., 1.],
        [2., 3.]], device='mps:0')
tensor([[0., 2.],
        [1., 3.]], device='mps:0')
```

Fixes introduced to `torch.mm` in https://github.com/pytorch/pytorch/pull/77462 suggests that this is not needed

Modify `sample_inputs_addmm` to test `torch.addmm` with square matrices, but skip this config for `test_autograd_dense_output_addmm`, see https://github.com/pytorch/pytorch/issues/116565


TODO: probably tweak tolerances, as `test_output_match_addmm_cpu_float16` fails with 2x2 matrices, but passes using 3x3 ones with errors slightly exceeding the tolerance

Fixes https://github.com/pytorch/pytorch/issues/116331

malfet committed:
[MPS] Fix addmm

Remove weird logic for designating matirces as transposed if sizes match
(which always true if square matrices are multiplied with each other)

Fixes introduced to `torch.mm` in
https://github.com/pytorch/pytorch/pull/77462 suggests that this is not
needed

Modify `sample_inputs_addmm` to test `torch.addmm` with square matrices

Fixes https://github.com/pytorch/pytorch/issues/116331

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Fix addmm"


Remove weird logic for designating matirces as transposed if sizes match
(which always true if square matrices are multiplied with each other)

Fixes introduced to `torch.mm` in
https://github.com/pytorch/pytorch/pull/77462 suggests that this is not
needed

Modify `sample_inputs_addmm` to test `torch.addmm` with square matrices

Fixes https://github.com/pytorch/pytorch/issues/116331

[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1145,8 +1145,8 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad, **kwargs):
     tests_with_lhs_broadcasting = [
-        ((1,), (2, 2), (2, 3), True),
-        ((), (2, 2), (2, 3), True)
+        ((1,), (2, 2), (2, 2), True),
```
Can you keep the old one as well?

malfet replied:
Sure, that was the original idea, because it doesn't even need broadcasting to exhibit the problem, just square matrices

albanD reviewed (approved):
Not sure what that logic was for indeed. Sounds good.

cc @kulinseth to take a look as well.

The sparse failures look real though, I would skip and open an issue.

malfet committed:
Update on "[MPS] Fix addmm"


Remove weird logic for designating matrices as transposed if sizes match(which always true if square matrices are multiplied with each other), which resulted in `torch.addmm` returns transposed matrix compared to `torch.mm`, see below:
```
python -c "import torch;torch.set_default_device('mps');a=torch.eye(2);b=torch.arange(4, dtype=torch.float).reshape(2, 2);print(ab);print(torch.addmm(torch.zeros(2, 2), a,b))"
tensor([[0., 1.],
        [2., 3.]], device='mps:0')
tensor([[0., 2.],
        [1., 3.]], device='mps:0')
```

Fixes introduced to `torch.mm` in https://github.com/pytorch/pytorch/pull/77462 suggests that this is not needed

Modify `sample_inputs_addmm` to test `torch.addmm` with square matrices

Fixes https://github.com/pytorch/pytorch/issues/116331

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet commented:
Ok, so float16 results are indeed slightly different between CPU and MPS, but I'm not sure if CPU has a proper implementation or whether MPS is wrong:
```
Greatest absolute difference: 0.000244140625 at index (1, 1) (up to 1e-05 allowed)
Greatest relative difference: 0.0010986328125 at index (1, 1) (up to 0.001 allowed)
```

malfet committed:
Update on "[MPS] Fix addmm"


Remove weird logic for designating matrices as transposed if sizes match(which always true if square matrices are multiplied with each other), which resulted in `torch.addmm` returns transposed matrix compared to `torch.mm`, see below:
```
% python -c "import torch;torch.set_default_device('mps');a=torch.eye(2);b=torch.arange(4.0).reshape(2, 2);print(ab);print(torch.addmm(torch.zeros(2, 2), a,b))"
tensor([[0., 1.],
        [2., 3.]], device='mps:0')
tensor([[0., 2.],
        [1., 3.]], device='mps:0')
```

Fixes introduced to `torch.mm` in https://github.com/pytorch/pytorch/pull/77462 suggests that this is not needed

Modify `sample_inputs_addmm` to test `torch.addmm` with square matrices, but skip this config for `test_autograd_dense_output_addmm`, see https://github.com/pytorch/pytorch/issues/116565

Fixes https://github.com/pytorch/pytorch/issues/116331

[ghstack-poisoned]

malfet commented:
@pytorchbot merge
